### ISSUE TITLE
Add whitespace control?

### ISSUE THREAD

**Issue by rduplain** [yogthos/Selmer#115]
Jinja has a useful feature where you can strip out whitespace that you've added to make your template more readable:

```jinja
{% for item in seq -%}
    {{ item }}
{%- endfor %}
```

> This will yield all elements without whitespace between them. If _seq_ was a list of numbers from `1` to `9`, the output would be `123456789`.

http://jinja.pocoo.org/docs/dev/templates/#whitespace-control

In other words, adding a `-` to get `-%}` will indicate that all whitespace should be removed between the `for`/`if`/other statement and the next character.

Feature request: any interest in adding this to Selmer?

**Comment by yogthos** [comment]
I'm not opposed to the feature, but I likely wouldn't have time to look at it in the near future. I could help with a pr if you're up for taking a look at it though.

**Comment by borkdude** [comment]
One obvervation made during a short experimental session:

The node structure looks like this (when printed from `render-template`:

```
user=> (selmer/render "{% if foo %} {{foo \n\n|json}} {% endif %}" {:foo 1})
:node #object[selmer.node.FunctionNode 0x648fa5b1 "selmer.node.FunctionNode@648fa5b1"]
:node #object[selmer.node.TextNode 0x74b5bb08 " "]
:node #object[selmer.node.FunctionNode 0x1da81d82 "selmer.node.FunctionNode@1da81d82"]
:node #object[selmer.node.TextNode 0x46d7ae7a " "]
:node #object[selmer.node.TextNode 0x151f9ad7 ""]
``` 

Maybe we can introduce a new type of node called a `SkipTextNode` which ignores text nodes (or only blank text nodes?) during rendering, until the next `SkipTextNode`.
And this `SkipTextNode` gets inserted during parsing after reading the `-%` sign?

Maybe I'm already thinking too much about the implementation, but it seems that this feature is possible to implement.

**Comment by yogthos** [comment]
Yeah I think we could just skip emitting these nodes during parsing.

**Comment by yogthos** [comment]
I think the relevant bit of code for this is [here](https://github.com/yogthos/Selmer/blob/bf88d83d135e71e2c3bc0eb2e705d4a280b1beac/src/selmer/util.clj#L129).

**Comment by borkdude** [comment]
I have a very hacky version here:

https://github.com/yogthos/Selmer/compare/master...borkdude:whitespace-control

``` clojure 
user=> (selmer/render "{% if foo %} \n hello \n {% endif %}" {:foo 1})
" \n hello \n "
user=> (selmer/render "{% if foo -%} \n hello \n {% endif %}" {:foo 1})
"hello"
```

I'm not sure if that's going into the right direction and also not sure if I'm a big fan of this magic syntax...

Would some other syntax work, like:

```
{% if foo }{% trim }

Yolo. All the whitespace you want!

{% endtrim}
{% endif %}
``` 

or is such a construct already possible, somehow?

**Comment by yogthos** [comment]
Adding a trim tag should work, but there's still a problem of the tags themselves leaving a blank line. With the way the parser works currently, there would be blank lines where the tags which I assume wouldn't be desirable.

**Comment by yogthos** [comment]
based on our slack brainstorming, let's go with `{% if(trim, x, y) foo %}` syntax where tag name could be followed by some arguments in parens akin to a function call. These could provide hints for how tag body would be further manipulated, such as trimming white space or empty lines.

**Comment by borkdude** [comment]
Adding this for posterity:

<img width="1350" alt="Screenshot 2021-05-10 at 21 23 11" src="https://user-images.githubusercontent.com/284934/117713443-f581b180-b1d5-11eb-94b0-ff1a895ab863.png">

/cc @roklenarcic

**Comment by RokLenarcic** [comment]
Any progress on this feature? The proposed solutions don't work because they control whitespace inside the tag, when the problem is the whitespace outside the tag. Here's a minimal example of a block of text that cannot be produced by the current Selmer version:

```
- Level: {{ level }}
- Color: {{ color }}
- Shape: {{ shape }}
```
Where level, color and shape are all optional, and if missing their line shouldn't be rendered and there must be no empty lines.

**Comment by Ramblurr** [comment]
Throwing another use case in here. Lines containing only selmer tags (e.g., `{% for thing in things %}`) end up as blank lines in the rendered output. For rendering plain-text (my use case) this is quite unfortunate. I'm not sure if either of the three jinja2 variants handle this case.

Btw, with the unlimited slack history.. here's the link to the thread screenshotted above: https://clojurians.slack.com/archives/C06MAR553/p1620329458137300

**Comment by yogthos** [comment]
This is a use case I'd be most interested in addressing as well.

**Comment by RokLenarcic** [comment]
This is close to 9 years open now...

**Comment by seancorfield** [comment]
Guess we're still waiting for someone to submit a PR...

**Comment by seancorfield** [comment]
I'll be honest, in all the years I've used Selmer this had never been a problem: I just put the `{` on different lines so there's no whitespace. In the few cases where whitespace really matters, this seems a reasonable compromise (making the template a little less reasonable).

**Comment by yogthos** [comment]
I haven't really found it to be a huge problem either, and implementing this efficiently is surprisingly tricky. I gave it a shot a few times and ended up giving up on it.

**Comment by RokLenarcic** [comment]
For me the main thing is not what initial post says (changing whitespace within the tag), but outside of it, e.g. I have multiple consecutive `if` blocks (or similar) and it's really hard to structure them in a way that doesn't leave loads of newlines behind if they evaluate to false. For me, 95% of the use of whitespace control would be to delete a single newline before or after the if, ifequals block.

**Comment by yogthos** [comment]
I'm definitely open to a PR for this if somebody wants to try give it a shot. The only tricky part here is that the parser needs to be able to peek ahead from the end of the tag to see if the rest of the line is empty or not. The default reader in Java only lets you scan a known umber of characters ahead as seen [here](https://github.com/yogthos/Selmer/blob/a535a5c55e078288e5da58b6be5e1ae1b744b3c1/src/selmer/util.clj#L118), so it would require a custom implementation that can keep reading into a buffer given a condition.
